File: output_repair/claude-opus/sample_3/repair_3/c/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

struct person
{
  char *name;
  struct person *spouse;
};

/*@ predicate person(struct person *p, struct person *sp) =
      p->name |-> _ &*& p->spouse |-> sp &*& malloc_block_person(p); @*/

struct person *create_person()
//@ requires true;
//@ ensures person(result, 0);
{
  struct person *p = malloc(sizeof(struct person));
  if (p == 0)
    abort();
  p->spouse = 0;

  return p;
}

void marry(struct person *this, struct person *other)
//@ requires person(this, _) &*& person(other, _);
//@ ensures person(this, other) &*& person(other, this);
{
  this->spouse = other;
  other->spouse = this;
}

struct person *person_get_spouse(struct person *this)
//@ requires person(this, ?s);
//@ ensures person(this, s) &*& result == s;
{
  return this->spouse;
}

void divorce(struct person *this)
//@ requires person(this, ?other) &*& other != 0 &*& person(other, this);
//@ ensures person(this, 0) &*& person(other, 0);
{
  this->spouse->spouse = 0;
  this->spouse = 0;
}

void die(struct person *this)
//@ requires person(this, 0);
//@ ensures true;
{
  if (this->spouse != 0)
  {
    this->spouse->spouse = 0;
  }
  free(this);
}

int main()
//@ requires true;
//@ ensures true;
{
  struct person *alice = create_person();
  struct person *bob = create_person();
  struct person *eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}

File: output_at_5/qwen3/sample_4/java/java/Linklist/Spouse.java
class Person {

    protected Person spouse;
    
    /*@
    predicate person(Person p) = true;
    @*/

    public  void spouse_symm()
        
        
    {
        //@ requires true;
        //@ ensures true;
        
    }

    public Person()
        
        
    {
        //@ requires true;
        //@ ensures this.spouse == null;
        
    }
    
    public Person getSpouse()
        
        
    {
        //@ requires true;
        //@ ensures result == this.spouse;
        
        return spouse;
        
    }
    
    protected void setSpouse(Person other)
        
        
    {
        //@ requires true &*& other != null;
        //@ ensures this.spouse == other &*& other.spouse == this;
        
        spouse = other;
        other.spouse = this;
        
    }
    
    protected void clearSpouse()
        
        
    {
        //@ requires true &*& spouse != null;
        //@ ensures this.spouse == null &*& spouse.spouse == null;
        
        spouse.spouse = null;
        spouse = null;
        
    }
    
    void marry(Person other)
        
        
    {
        //@ requires true &*& other != null;
        //@ ensures true;
        
        other.setSpouse(this);
        
    }
    
    void divorce()
        
        
    {
        //@ requires true &*& spouse != null;
        //@ ensures this.spouse == null;
        
        spouse.clearSpouse();
        
    }

}

class Program {

    public static void main(String[] args)
        
        
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}